### PR TITLE
[Feature] Add torch_npu optimizer

### PR DESCRIPTION
**Pull request by luomaoling**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add npu optimizer.

## Modification

mmengine/optim/optimizer/builder.py.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
None.

## Use cases (Optional)

python tools/train.py configs/resnet/resnet50_8xb32_in1k.py --cfg-options optimizer.type=NpuFusedSGD.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmengine/optim/optimizer/builder.py [hunk 1]
```diff
@@ -33,6 +33,32 @@ def register_torch_optimizers() -> List[str]:
 TORCH_OPTIMIZERS = register_torch_optimizers()
 
 
+def register_torch_npu_optimizers() -> List:
```
Comment (HAOCHENYE) [inline mmengine/optim/optimizer/builder.py:36]:
```suggestion
def register_torch_npu_optimizers() -> List[str]:
```

**Review by HAOCHENYE — DISMISSED** [review]

File: mmengine/optim/optimizer/builder.py [hunk 2]
```diff
@@ -33,6 +33,36 @@ def register_torch_optimizers() -> List[str]:
...
+    already_registered_optim = OPTIMIZERS.__dict__['_module_dict'].keys()
+    for module_name in dir(torch_npu.optim):
+        if module_name.startswith(
+                '__') or module_name in already_registered_optim:
+            continue
+        _optim = getattr(torch_npu.optim, module_name)
+        if inspect.isclass(_optim) and issubclass(_optim,
+                                                  torch.optim.Optimizer):
+            OPTIMIZERS.register_module()(_optim)
+            torch_npu_optimizers.append(module_name)
```
Comment (zhouzaida) [inline mmengine/optim/optimizer/builder.py:59]:
```suggestion
    for module_name in dir(torch_npu.optim):
        if module_name.startswith('__') or module_name in OPTIMIZERS:
            continue
        _optim = getattr(torch_npu.optim, module_name)
        if inspect.isclass(_optim) and issubclass(_optim,
                                                  torch.optim.Optimizer):
            OPTIMIZERS.register_module(module=_optim)
            torch_npu_optimizers.append(module_name)
```

Comment (zhouzaida) [reply mmengine/optim/optimizer/builder.py:59]: `Registry` implemented the `__contains__` method, so we can directly use `module_name in OPTIMIZERS` to check whether an item exists.

File: docs/zh_cn/common_usage/speed_up_training.md [hunk 3]
```diff
@@ -114,3 +114,21 @@ runner = Runner(
 ```{warning}
 `torch.compile` 目前仍然由 PyTorch 团队持续开发中，一些模型可能会编译失败。如果遇到了类似问题，你可以查阅 [PyTorch Dynamo FAQ](https://pytorch.org/docs/2.0/dynamo/faq.html) 解决常见问题，或参考 [TorchDynamo Troubleshooting](https://pytorch.org/docs/2.0/dynamo/troubleshooting.html) 向 PyTorch 提 issue.
 ```
+
+## 优化器
```
Comment (zhouzaida) [inline docs/zh_cn/common_usage/speed_up_training.md:118]:
```suggestion
## 使用更快的优化器
```

File: docs/zh_cn/common_usage/speed_up_training.md [hunk 4]
```diff
@@ -114,3 +114,21 @@ runner = Runner(
...
+NpuFusedAdadelta
+NpuFusedAdam
+NpuFusedAdamP
+NpuFusedAdamW
+NpuFusedBertAdam
+NpuFusedLamb
+NpuFusedRMSprop
+NpuFusedRMSpropTF
+NpuFusedSGD
+```
```
Comment (zhouzaida) [inline docs/zh_cn/common_usage/speed_up_training.md:132]:
```suggestion

- NpuFusedAdadelta
- NpuFusedAdam
- NpuFusedAdamP
- NpuFusedAdamW
- NpuFusedBertAdam
- NpuFusedLamb
- NpuFusedRMSprop
- NpuFusedRMSpropTF
- NpuFusedSGD

```

File: docs/en/common_usage/speed_up_training.md [hunk 5]
```diff
@@ -113,3 +113,21 @@ This feature is only available for PyTorch >= 2.0.0.
...
+NpuFusedAdadelta
+NpuFusedAdam
+NpuFusedAdamP
+NpuFusedAdamW
+NpuFusedBertAdam
+NpuFusedLamb
+NpuFusedRMSprop
+NpuFusedRMSpropTF
+NpuFusedSGD
+```
```
Comment (zhouzaida) [inline docs/en/common_usage/speed_up_training.md:131]:
```suggestion

- NpuFusedAdadelta
- NpuFusedAdam
- NpuFusedAdamP
- NpuFusedAdamW
- NpuFusedBertAdam
- NpuFusedLamb
- NpuFusedRMSprop
- NpuFusedRMSpropTF
- NpuFusedSGD

```

File: docs/zh_cn/common_usage/speed_up_training.md [hunk 6]
```diff
@@ -114,3 +114,21 @@ runner = Runner(
...
+- NpuFusedAdamP
+- NpuFusedAdamW
+- NpuFusedBertAdam
+- NpuFusedLamb
+- NpuFusedRMSprop
+- NpuFusedRMSpropTF
+- NpuFusedSGD
+
+
+使用方式同原生优化器一样，MMEngine 中更详细的使用方式可参考 [optimizers](https://mmengine.readthedocs.io/zh_CN/latest/tutorials/optim_wrapper.html?highlight=%E4%BC%98%E5%8C%96%E5%99%A8).
```
Comment (zhouzaida) [inline docs/zh_cn/common_usage/speed_up_training.md:134]:
```suggestion
如果使用了昇腾的设备，可以使用昇腾的优化器从而缩短模型的训练时间。昇腾设备支持的优化器如下

- NpuFusedAdadelta
- NpuFusedAdam
- NpuFusedAdamP
- NpuFusedAdamW
- NpuFusedBertAdam
- NpuFusedLamb
- NpuFusedRMSprop
- NpuFusedRMSpropTF
- NpuFusedSGD

使用方式同原生优化器一样，可参考[优化器的使用](../tutorials/optim_wrapper.md#在执行器中配置优化器封装)。
```

File: docs/en/common_usage/speed_up_training.md [hunk 7]
```diff
@@ -113,3 +113,21 @@ This feature is only available for PyTorch >= 2.0.0.
...
+
+The usage method is the same as the native optimizer, and the more detailed usage method in MMEngine can be referred to [optimizers](https://mmengine.readthedocs.io/zh_CN/latest/tutorials/optim_wrapper.html?highlight=%E4%BC%98%E5%8C%96%E5%99%A8).
```
Comment (zhouzaida) [inline docs/en/common_usage/speed_up_training.md:133]: Please also update the en doc.

File: docs/en/common_usage/speed_up_training.md [hunk 8]
```diff
@@ -113,3 +113,19 @@ This feature is only available for PyTorch >= 2.0.0.
...
+- NpuFusedAdam
+- NpuFusedAdamP
+- NpuFusedAdamW
+- NpuFusedBertAdam
+- NpuFusedLamb
+- NpuFusedRMSprop
+- NpuFusedRMSpropTF
+- NpuFusedSGD
+
+The usage method is the same as the native optimizer, please refer to [use of optimizers](../tutorials/optim_wrapper.md#configure-the-optimwapper-in-runner)
```
Comment (zhouzaida) [inline docs/en/common_usage/speed_up_training.md:131]:
```suggestion
## Using faster Optimizers

If Ascend devices are used, you can use the Ascend optimizers to shorten the training time of the model. The optimizers supported by Ascend devices are as follows:

- NpuFusedAdadelta
- NpuFusedAdam
- NpuFusedAdamP
- NpuFusedAdamW
- NpuFusedBertAdam
- NpuFusedLamb
- NpuFusedRMSprop
- NpuFusedRMSpropTF
- NpuFusedSGD

The usage is the same as native optimizers, and you can refer to [Using Optimizers](../tutorials/optim_wrapper.md#configure-the-optimwapper-in-runner) for more information.
```

**Review by zhouzaida — APPROVED** [review]

**Pull request merged**